matrix_benchmarking/regression.py

```diff
@@ -0,0 +1,176 @@
+            combine_funcs: dict = {},
+        ):
+        self.new_payload = new_payload
+        self.x_var = x_var
+        self.x_var_key = x_var_key
+        self.kpis = kpis
+        self.combine_funcs = combine_funcs
+        self.settings_filter = settings_filter
+
+        if self.settings_filter and self.x_var:
+            logging.warning("settings_filter and x_var set, only using settings_filter")
+        elif self.x_var:
+            settings = self.new_payload.get_settings()
+            settings.pop(self.x_var)
+            self.settings_filter = settings
+
+        if self.settings_filter:
+            # Only store payloads that have equivalent (k, v) pairs
+            # as the settings_filter
+            self.lts_payloads = list(
+                filter(
+                    lambda x: dict_part_eq(self.settings_filter, x.get_settings()),
+                    lts_payloads
+                )
+            )
+
+            if not dict_part_eq(self.settings_filter, self.new_payload.get_settings()):
+                self.new_payload = None
+                logging.warning("settings_filter isn't satisfied for the new payload")
+        else:
+            self.lts_payloads = lts_payloads
+
+        # This isn't strictly necessary for all analysis techniques, but
+        # is useful to have
+        self.lts_payloads.sort(key=lambda entry: self.x_var_key(entry))
+
+
+    def analyze(self) -> list[dict]:
+
+        if not self.new_payload:
+            return [{"result": None, "kpi": None, "regression": vars(RegressionStatus(0, explanation="Not enough new data"))}]
+
+        if not self.lts_payloads:
+            return [{"result": None, "kpi": None, "regression": vars(RegressionStatus(0, explanation="Not enough LTS data"))}]
+
+        regression_results = []
+
+        kpis_to_test = vars(self.new_payload.results.lts.kpis).keys() if not self.kpis else self.kpis
+        for kpi in kpis_to_test:
+
+            curr_values = vars(self.new_payload.results.lts.kpis)[kpi].value
+            lts_values = list(map(lambda x: vars(x.results.kpis)[kpi].value, self.lts_payloads))
+
+            if type(vars(self.new_payload.results.lts.kpis)[kpi].value) is list:
+                if kpi in self.combine_funcs:
+                    curr_values = self.combine_funcs[kpi](curr_values)
+                    lts_values = [self.combine_funcs[kpi](v) for v in lts_values]
+                else:
+                    logging.warning(f"Skipping KPI with list of values, consider filtering KPIs or providing a combine_func for {kpi}")
+                    continue
+
+            regression_results.append(
+                {
+                    "result": self.new_payload.get_settings(),
+                    "kpi": kpi,
+                    "regression": vars(
+                        self.regression_test(
+                            curr_values,
+                            lts_values
+                        )
```

Comment:
can you save the result of this call into a variable, and use the variable here?
this is a key the important part of the function body, shouldn't be "hidden" in the middle of the object creation

Reply:
Sure I'll make those more clear

Reply:
@kpouget How do you mean: `and everything in __init__.py`?

Reply:
in the `regression` directory, you'll create some modules/files, 
and what does not belong to a module, you put it in `regression/__init__.py`, so that it's reachable with eg `regression.RegressionStatus(`